core/src/main/java/software/amazon/awssdk/interceptor/context/BeforeExecutionContext.java

```diff
@@ -0,0 +1,33 @@
+/*
+ * Copyright 2010-2017 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package software.amazon.awssdk.interceptor.context;
+
+import software.amazon.awssdk.SdkRequest;
+import software.amazon.awssdk.annotation.SdkPublicApi;
+import software.amazon.awssdk.annotation.ThreadSafe;
+import software.amazon.awssdk.interceptor.ExecutionInterceptor;
+
+/**
+ * The state of the execution when the {@link ExecutionInterceptor#beforeExecution} method is invoked.
+ */
+@ThreadSafe
+@SdkPublicApi
+public interface BeforeExecutionContext {
```

Comment:
It feels a bit overkill since people will probably be learning about them from the interceptor javadoc or any developer docs we end up writing on them. It also makes imports a pain - some IDEs include the parent class by default. It's also not very common to do something like that in Java, from all that I have seen.